### ISSUE TITLE
chore(pytrees): migrate constraints and frozen dataclasses to equinox.Module

### ISSUE THREAD

**Issue by antonioterpin** [antonioterpin/pinet#98]
## Motivation

We currently maintain two parallel conventions for "things that carry JAX arrays and cross `jit` boundaries":

1. **Frozen dataclasses registered as pytrees by hand** in [src/pinet/dataclasses.py](src/pinet/dataclasses.py). Every class (`EqualityConstraintsSpecification`, `BoxConstraintSpecification`, `SocConstraintSpecification`, `NonLinearSpecification`, `ProjectionInstance`, `EquilibrationParams`) needs the `@jax.tree_util.register_dataclass` + `@dataclass(frozen=True)` pair, a hand-written `update(**kwargs)` delegating to `replace`, and a custom `validate()` that walks attributes. `NonLinearSpecification` additionally has to spell out `data_fields` / `meta_fields` by hand ([dataclasses.py:304-308](src/pinet/dataclasses.py#L304-L308)) because `nl_type` is a non-array field.
2. **Plain Python classes holding arrays and bound methods** in `src/pinet/constraints/*.py` (`EqualityConstraint`, `AffineInequalityConstraint`, `BoxConstraint`, `SocConstraint`, `CartesianConstraint`, `NonLinearConstraint`). These are captured by closure inside jitted code — e.g. [project.py:241-252](src/pinet/project.py#L241-L252) and [solver/admm.py:82-143](src/pinet/solver/admm.py#L82-L143) rely on `self` being baked into a traced closure. Swapping array parameters on a constraint therefore triggers a retrace.

Both patterns are what [equinox](https://docs.kidger.site/equinox/) was built for. `equinox.Module` gives us pytree registration, frozen-by-default dataclass semantics, static-vs-dynamic field discrimination (`eqx.field(static=True)`), and bound methods that are themselves pytree-valued — no closure-capture needed.

### Step 1: replace hand-registered dataclasses with `eqx.Module`

Target files: [src/pinet/dataclasses.py](src/pinet/dataclasses.py), [src/pinet/equilibration.py](src/pinet/equilibration.py) (consumes `EquilibrationParams`).

```python
class EqualityConstraintsSpecification(eqx.Module):
    b:          Float[Array, "B m 1"] | None = None
    a_dyn:      Float[Array, "B m n"] | None = None
    a_dyn_pinv: Float[Array, "B n m"] | None = None
```

- Drops `@jax.tree_util.register_dataclass`, `@dataclass(frozen=True)`, the per-class `update(**kwargs)` helper (`eqx.tree_at` covers the niche cases; field assignment via `__init__` covers the rest), and the `data_fields`/`meta_fields` split for `NonLinearSpecification` (use `eqx.field(static=True)` on `nl_type`).
- `validate()` methods stay as plain methods — no framework change needed.

### Step 2: migrate `Constraint` classes to `eqx.Module`

Target files: [src/pinet/constraints/base.py](src/pinet/constraints/base.py) plus every subclass in `src/pinet/constraints/`.

- `Constraint` becomes an abstract `eqx.Module`.
- Array attributes (`a_dyn`, `b`, `lb`, `ub`, `mask`, …) become dynamic fields; scalar/string configuration (`method`, `var_b`, `var_a_dyn`, shape-only constants like `EXPECTED_NDIM`) becomes static via `eqx.field(static=True)`.
- The trickiest case is `EqualityConstraint` — `setup()` currently rebinds `self.project` to one of `project_pinv` or a raiser depending on `method` ([affine_equality.py:95-107](src/pinet/constraints/affine_equality.py#L95-L107)). With `eqx.Module` we instead dispatch on the static `method` field inside a single `project` method; this is both cleaner and trace-stable.
- Same pattern applies in [soc_constraint.py](src/pinet/constraints/soc_constraint.py), [box.py](src/pinet/constraints/box.py), [cartesian_constraint.py](src/pinet/constraints/cartesian_constraint.py).

### Step 3: simplify the `Project` jit boundary

In [project.py:241-252](src/pinet/project.py#L241-L252) we currently pre-bake closures and hand them to `jax.jit(partial(..., static_argnames=["n_iter", "n_iter_bwd", "fpi"]))`. Once constraints are `eqx.Module`s, the natural replacement is:

- Use `eqx.filter_jit` at the boundary so the array leaves trace and the static parts (including bound methods) are automatically held static. This removes the `partial` gymnastics and the need to enumerate `static_argnames` by hand.
- For the `jax.custom_vjp` path in [project.py:441-482](src/pinet/project.py#L441-L482), `nondiff_argnames` stays — but the Callable arguments can be wrapped in `eqx.Module` so the partition/combine boundary is explicit and discoverable. If the implicit approach causes friction, drop to `eqx.partition` / `eqx.combine` inside the forward/backward as recommended by the equinox docs.

## Why this is worth the churn

- Removes ~6 custom `update()` methods plus the `register_dataclass` boilerplate.
- Makes "what is static vs traced" an attribute of the class, not something inferred from which closure captured what.
- Lets us pass constraint instances directly through `vmap`/`scan`/`custom_vjp` without closure trickery — useful the moment we want per-batch constraint parameters, which [issue #50](https://github.com/antonioterpin/pinet/issues/50) already foreshadows.
- Aligns with the project's JAX-first stance from [docs/standards/ml-numerical.md](docs/standards/ml-numerical.md).

## Suggested order

- [ ] Add `equinox` as a runtime dep (`uv add equinox`).
- [ ] Migrate `src/pinet/dataclasses.py` (leaf types — lowest blast radius).
- [ ] Migrate `Constraint` subclasses one file at a time, keeping the public API identical.
- [ ] Flip `Project._project` to `eqx.filter_jit` and drop the `partial`/`static_argnames` dance.
- [ ] Evaluate whether the `custom_vjp` path benefits from explicit `eqx.partition` / `eqx.combine` or stays as-is.

**Comment by antonioterpin** [comment]
Landed via #104 (equinox.Module migration of constraints + frozen dataclasses).
